### ISSUE TITLE
CSP headers are incorrect with multiple rules

### ISSUE THREAD

**Issue by letanloc1998** [w3c/webappsec-csp#682]
Hi everyone,

When we config CSP header like:

```
Content-Security-Policy: a
Content-Security-Policy: b
Content-Security-Policy: c
```

Each header is independent. (It means we must write rule for both script-src-elem  and connect-src,...)

How to split Content-Security-Policy to multiple headers? Because it's very long.

Many thanks

**Comment by mikewest** [comment]
I don't think I understand the question.

```http
Content-Security-Policy: a
Content-Security-Policy: b
Content-Security-Policy: c
```

would imply that `a` and `b` and `c` are all enforced. You don't _need_ to specify each directive in a distinct header, though:

```http
Content-Security-Policy: a; b; c
```

would also enforce `a`, `b`, and `c`. The two models differ in some edge cases that are important (e.g. enforcing hashes or nonces _and_ an allowlist of domains), but they're often interchangable.

Is that somewhere near what you're asking?

**Comment by mikewest** [comment]
Yes. But `Content-Security-Policy: a, b, c` is not identical to `Content-Security-Policy: a; b; c`. The former creates three policies, all of which are enforced. The latter creates one policy with three directives, all of which are enforced. These are distinct. For example, a simplified version of the headers sent from `myaccount.google.com` is:

```http
content-security-policy: script-src ... 'nonce-75mShyMY4JEfhPfGuk2DeA';object-src 'none'; ...
content-security-policy: script-src https://apis.google.com [more domains go here]; ...
```

That enforces a nonce constraint and also enforces an allowlist of domains. That's not possible to do in a single policy.

**Comment by dveditz** [comment]
> How to split Content-Security-Policy to multiple headers? Because it's very long.

We've talked about ways to specify CSP in a separately-loaded manifest, or allowing CSP to be set by Origin Policy or Document Policy if those ever get specified and implemented—but so far that has only been talk. For now you don't have many good options. Simplify your site? Decide some types of resources aren't worth restricting?

If you do want to try to split it you have to be careful to only use independent directives. For example, you couldn't specify `default-src` in any policy because that will end up being enforced for any missing `-src` directive, which will then conflict with the setting for any directive that was moved into a separate policy.

See the lists at https://w3c.github.io/webappsec-csp/#directive-fallback-list

You can split different directives into separate headers as long as you never put any of the directives that are on the same line into different policy headers. Ideally you would only use the left-most directives and none of the ones on the right -- and especially not `default-src`.
